src/deim.jl

```diff
@@ -0,0 +1,167 @@
+    g = constant_part
+    F = nonlinear_part
+    ŷ = reduced_vars
+    V = linear_projection_matrix
+    U = nonlinear_projection_matrix
+
+    # original vars to reduced vars
+    linear_projection_eqs = Symbolics.scalarize(y .~ V * ŷ)
+    linear_projection_dict = Dict(eq.lhs => eq.rhs for eq in linear_projection_eqs)
+
+    indices = deim_interpolation_indices(U) # DEIM interpolation indices
+    # the DEIM projector (not DEIM basis) satisfies
+    # F(original_vars) ≈ projector * F(pod_basis * reduced_vars)[indices]
+    projector = ((@view U[indices, :])' \ (U' * V))'
+    temp = substitute.(F[indices], (linear_projection_dict,))
+    F̂ = projector * temp # DEIM approximation for nonlinear func F
+
+    Â = V' * A * V
+    ĝ = V' * g
+    reduced_rhss = Â * ŷ + ĝ + F̂
+    reduced_rhss, linear_projection_eqs
+end
+"""
+$(TYPEDSIGNATURES)
+
+Reduce a `ModelingToolkit.ODESystem` using the Proper Orthogonal Decomposition (POD) with
+the Discrete Empirical Interpolation Method (DEIM).
+
+`snapshot` should be a matrix with the data of each time instance as a column.
+
+The LHS of equations in `sys` are all assumed to be 1st order derivatives. Use
+`ModelingToolkit.ode_order_lowering` to transform higher order ODEs before applying DEIM.
+
+`sys` is assumed to have no internal systems. End users are encouraged to call
+`ModelingToolkit.structural_simplify` beforehand.
+
+The POD basis used for DEIM interpolation is obtained from the snapshot matrix of the
+nonlinear terms, which is computed by executing the runtime-generated function for
+nonlinear expressions.
+"""
+function deim(sys::ODESystem, snapshot::AbstractMatrix, pod_dim::Integer;
+              deim_dim::Integer = pod_dim,
+              name::Symbol = Symbol(nameof(sys), :_deim))::ODESystem
+    @set! sys.name = name
+
+    # handle ODESystem.substitutions
+    # https://github.com/SciML/ModelingToolkit.jl/issues/1754
+    sys = tearing_substitution(sys)
+
+    iv = ModelingToolkit.get_iv(sys) # the single independent variable
+    D = Differential(iv)
+    dvs = ModelingToolkit.get_states(sys) # dependent variables
+
+    pod_reducer = POD(snapshot, pod_dim)
+    reduce!(pod_reducer, TSVD())
+    V = pod_reducer.rbasis # POD basis
+
+    var_name = gensym(:ŷ)
+    ŷ = (@variables $var_name(iv)[1:pod_dim])[1]
+    @set! sys.states = Symbolics.value.(Symbolics.scalarize(ŷ)) # new variables from POD
+    ModelingToolkit.get_var_to_name(sys)[Symbolics.getname(ŷ)] = Symbolics.unwrap(ŷ)
+
+    deqs, eqs = get_deqs(sys) # split eqs into differential and non-differential equations
+    rhs = Symbolics.rhss(deqs)
+    # a sparse matrix of coefficients for the linear part,
+    # a vector of constant terms and a vector of nonlinear terms about dvs
+    A, g, F = linear_terms(rhs, dvs)
+
+    # generate an in-place function from the symbolic expression of the nonlinear functions
+    F_expr = build_function(F, dvs; expression = Val{false})[2]
+    F_func! = eval(F_expr)
```

Comment:
This makes it not safe to use in a function. Is there an issue with RuntimeGeneratedFunctions here?

Reply:
I haven't gotten an issue here inside `deim`, but I did get errors when solving large systems as described in 
- JuliaSymbolics/Symbolics.jl#738

Reply:
I see. I updated this in commit https://github.com/SciML/ModelOrderReduction.jl/pull/19/commits/34645814a2b2ff361c7073082de0e300ddff1486 in which `eval` in removed. Now with the option `expression = Val{false}`, `Symbolics.build_function` internally utilizes RuntimeGeneratedFunctions.jl to be free of world-age issues.

Reply:
👍